src/ophys_etl/modules/segmentation/graph_utils/feature_vector_rois.py

```diff
@@ -0,0 +1,522 @@
+        An (n_rows, n_cols) array of booleans marked True at any pixels
+        that should be ignored, presumably because they have already been
+        selected as ROI pixels. (default: None)
+
+    Returns
+    -------
+    features: np.ndarray
+         An (n_rows*n_cols, n_rows*n_cols) array of feature vectors.
+         features[ii, :] is the feature vector for the ii_th pixel
+         in sub_video (where ii can be mapped to (row, col) coordinates
+         using np.unravel_index()
+
+    Notes
+    -----
+    The feature vectors returned by this method are calculated as follows:
+
+    1) Select a random set of 10 pixels from sub_video including the seed_pt.
+
+    2) Use the brightest filter_fraction of timestamps from the 10 selected
+    seeds to calculate the Pearson correlation coefficient between each
+    pair of pixels. This results in an (n_pixels, n_pixels) array in which
+    pearson[ii, jj] is the correlation coefficient between the ii_th and
+    jj_th pixels.
+
+    3) For each column pearson [:, ii] in the array of Pearson
+    correlation coefficients, subtract off the minimum value in
+    that column and divide by the gap between the 25th adn 75th
+    percentile of that column's values.
+
+    Each row in the matrix of Pearson correlation coefficients is now
+    a feature vector corresponding to that pixel in sub_video.
+    """
+
+    # fraction of timesteps to discard
+    discard = 1.0-filter_fraction
+
+    n_rows = sub_video.shape[1]
+    n_cols = sub_video.shape[2]
+
+    # start assembling mask in timesteps
+    trace = sub_video[:, seed_pt[0], seed_pt[1]]
+    thresh = np.quantile(trace, discard)
+    global_mask = []
+    mask = np.where(trace >= thresh)[0]
+    global_mask.append(mask)
+
+    # choose n_seeds other points to populate global_mask
+    i_seed = np.ravel_multi_index(seed_pt, sub_video.shape[1:])
+    possible_seeds = []
+    for ii in range(n_rows*n_cols):
+        if ii == i_seed:
+            continue
+        p = np.unravel_index(ii, sub_video.shape[1:])
+
+        if pixel_ignore is None or not pixel_ignore[p[0], p[1]]:
+            possible_seeds.append(ii)
+
+    n_seeds = 10
+    if rng is None:
+        rng = np.random.RandomState(87123)
+    chosen = set()
+    chosen.add(seed_pt)
+
+    if len(possible_seeds) > n_seeds:
+        chosen_dex = rng.choice(possible_seeds,
+                                size=n_seeds, replace=False)
+        for ii in chosen_dex:
+            p = np.unravel_index(ii, sub_video.shape[1:])
+            chosen.add(p)
+    else:
+        for ii in possible_seeds:
+            p = np.unravel_index(ii, sub_video.shape[1:])
+            chosen.add(p)
+
+    for chosen_pixel in chosen:
+        trace = sub_video[:, chosen_pixel[0], chosen_pixel[1]]
+        thresh = np.quantile(trace, discard)
+        mask = np.where(trace >= thresh)[0]
+        global_mask.append(mask)
+    global_mask = np.unique(np.concatenate(global_mask))
```

Comment:
so, it seems like you've slipped back into a global neighborhood mask, while, in the original filtered hnc metric exploration, you were seeing improved contrast with pairwise masks. Can you explain?

Reply:
My initial thought was this:

When generating the "image" graphs, we are only caring about how two neighboring pixels relate to each other, so you can ask "what are the brightest timestamps for pixel0 and pixel1." Here, we are correlating the entire neighborhood of 1600 (40 x 40) pixels, so there aren't just two pixels. Put another way, I wanted features[ii, jj] and features[ii+10, jj] to be calculated with the same timesteps.

That being said...

I could, for each pair of pixels, find the brightest timesteps. That means that the feature vectors would be a little inhomogeneous. features[ii, jj] and features[ii+10, jj] would not be calculated with the same timesteps. Maybe that doesn't matter.

I am worried about implementation. If, for every pair of pixels, we use a different set of timesteps, that is 1600*1599/2 different correlations that cannot easily be run through numpy.tensordot. Originally, tensordot took me down from processing an ROI in 5 minutes to processing an ROI in 5 seconds. A lot has changed since that timing test, though, so maybe there is a way to make it work. I will try it. As you suggest: the results might be much better. I will let you know.

Reply:
I think it is worth trying out. I see how it might slow things down